ts/src/flexible-event/privacy.ts

```diff
@@ -123,46 +133,73 @@ export function flipProbabilityDp(numStates: number, epsilon: number): number {
   return numStates / (numStates + Math.exp(epsilon) - 1)
 }
 
-/**
- * Computes the capacity of the q-ary symmetric channel.
- *
- * @param log2q - the logarithm to base 2 of the alphabet size.
- * @param flipProbability - the channel keeps the input the same with probability
- *   1 - flipProbability, and flips the input to one of the other q - 1
- *   symbols (uniformly) with the remaining probability of flipProbability.
- *
- * @returns - The capacity of the q-ary symmetric channel for given flipProbability.
- *   In general, the capacity is defined as the maximum, over all input
- *   distributions, of the mutual information between the input and output of
- *   the channel. In the special case of the q-ary symmetric channel, a
- *   closed-form expression is known, which we use here.
- */
-function capacityQarySymmetricChannel(
-  log2q: number,
-  flipProbability: number
+// Returns the pick rate for a source with attribution scope.
+export function fakeProbabilityDp(
+  numStates: number,
+  numUsedScopes: number,
+  numEventStates: number,
+  epsilon: number
 ): number {
+  const pickRateForSource =
+    (flipProbabilityDp(numStates, epsilon) * (numStates - 1)) / numStates
+  const pickRateForEvent =
+    (flipProbabilityDp(numEventStates, epsilon) * (numEventStates - 1)) /
+    numEventStates
   return (
-    log2q -
-    binaryEntropy(flipProbability) -
-    flipProbability * Math.log2(Math.pow(2, log2q) - 1)
+    1 - (1 - pickRateForSource) * Math.pow(1 - pickRateForEvent, numUsedScopes)
   )
 }
 
-export function maxInformationGain(numStates: number, epsilon: number): number {
-  const flipProb = flipProbabilityDp(numStates, epsilon)
-  if (numStates === 1 || flipProb === 1) {
+export function informationGain(
+  numStates: number,
+  numUsedScopes: number,
+  numEventStates: number,
+  epsilon: number
+): number {
+  // Use BigInt for the total number of states to avoid potential overflow.
+  const totalNumStates =
+    BigInt(numStates) + BigInt(numEventStates) * BigInt(numUsedScopes)
+  if (totalNumStates <= 1n) {
     return 0
   }
-  return capacityQarySymmetricChannel(
-    Math.log2(numStates),
-    (flipProb * (numStates - 1)) / numStates
+  const log2Q = Math.log2(Number(totalNumStates))
```

Comment:
Is there a problem here if `totalNumStates` can't be precisely represented as a `number`?

Reply:
Actually I don't think it's necessary to convert the numbers to BigInt as `number` can represent float and integers in type script, and no overflow issues if the multiplication gets too big.